## Github Repository for
#   A genomic resource for exploring bacterial-viral dynamics in seagrass ecosystems

## by Cassie Ettinger

### To cite this work or code

<i> Preprint coming soon </i>

### Abstract

Background: Seagrasses are globally distributed marine flowering plants that play foundational roles in coastal environments as ecosystem engineers. While research efforts have explored various aspects of seagrass-associated microbial communities, including describing the diversity of bacteria, fungi and microbial eukaryotes, little is known about viral diversity in these communities. 

Results: To begin to address this, we leveraged metagenomic sequencing data to generate a catalog of bacterial metagenome-assembled genomes (MAGs) and phage genomes from the leaves of the seagrass, <i>Zostera marina</i>. We expanded the robustness of this viral catalog by incorporating publicly available metagenomic data from seagrass ecosystems. The final MAG set represents 85 high-quality draft and 62 medium-quality draft bacterial genomes. While the viral catalog represents 354 medium-quality, high-quality, and complete viral genomes. Predicted auxiliary metabolic genes in the final viral catalog had putative annotations largely related to carbon utilization, suggesting a possible role for phage in carbon cycling in seagrass ecosystems. 

Conclusions: These genomic resources provide initial insight into bacterial-viral interactions in seagrass meadows, and are a foundation on which to further explore these critical interkingdom interactions. These catalogs highlight a possible role for viruses in carbon cycling in seagrass beds which may have important implications for blue carbon management and climate change mitigation.



 

